### PR TITLE
Optimized retainAll removeIf and forEach for array backed List and Set

### PR DESCRIPTION
**Pull request by shishu94**
Added optimized methods "retainAll", "removeIf" and "forEach" based on the optimized method
"removeAll" for arrays.
Added test to cover those 4 methods in the IntArrayListTest and IntArraySetTest

### CODE REVIEW

**Comment by vigna** [comment]
Did you benchmark the two implementations? I have no idea of the overhead of having contains passed by pointer.

**Comment by shishu94** [comment]
Hello, this was a long time ago :)
At that time i micro benched with JMH this code and a version with specialized code for each function and if i remember correctly the performance was the same (I simply preferred this cleaner version).
But if you need some evidences of this, i can write the bench again.

**Comment by vigna** [comment]
No, it's OK. Are the unit tests covering all the code? And I'll need your real name if you wanna be thanked in the CHANGES.

File: drv/ArrayList.drv [hunk 1]
```diff
@@ -862,7 +892,7 @@ public class ARRAY_LIST KEY_GENERIC extends ABSTRACT_LIST KEY_GENERIC implements
 			catch (IndexOutOfBoundsException e) { mThrowsOutOfBounds = e; }
 
 			try {
-				t.add(p, KEY2OBJ(T));
+				t.add(p, KEY2OBJ(T)"""""");
```
Comment (shishu94) [inline drv/ArrayList.drv]: I think my keyboard got mad here just before doing the commit

**Comment by shishu94** [comment]
Yes the tests cover each case of the added methods

**Comment by vigna** [comment]
OK, please confirm you don't want your name in.

**Comment by shishu94** [comment]
You can add my name. It is DOS SANTOS JERONIMO Leonel
Do you need me to push a correction for the comment i did or can you fix it ?

**Comment by vigna** [comment]
Ouch, ReferenceArraySet does not compile :(.

**Comment by shishu94** [comment]
Ok, so i'll try to setup the build today and check what is happening

**Comment by vigna** [comment]
Argh, I think I've been too quick. There are numerous problem with this pull request. Here

 return removeIf((java.util.function.IntPredicate) c::contains);

the method is deprecated and Eclipse warns that the conversion is unlikely. Does this work?

**Comment by vigna** [comment]
In general, i have warnings for boxing/unboxing, deprecation, etc. and I expect the code to be completely clean of those.

**Comment by vigna** [comment]
I fixed part of the problems (unqualified Arrays calls, missing cast for arrays) but I'm a bit wary at this point.

**Comment by shishu94** [comment]
No problem, as i said since this has been made some time ago, if you suspect something, i will check the build and rework the pull request if needed. I don't have my laptop now, as soon as i can i'll do it.

**Comment by shishu94** [comment]
If the problem if only on c::contains my wild guess is that eclipse does not like to transform the method reference as an IntPredicate. Replacing c::contains with "v -> c.contains(v)" should be fine.

**Pull request merged**